wireless: T6425: Fixing VHT beamforming for 802.11ac

alainlamar opened the pull request:
<!-- All PR should follow this template to allow a clean and transparent review -->
<!-- Text placed between these delimiters is considered a comment and is not rendered -->

## Change Summary
<!--- Provide a general summary of your changes in the Title above -->
* VHT beamforming support was broken which was fixed. 
* Smoketests were added to detect new bugs on VHT beamforming.

## Types of changes
<!---
What types of changes does your code introduce? Put an 'x' in all the boxes that apply.
NOTE: Markdown requires no leading or trailing whitespace inside the [ ] for checking
the box, please use [x]
-->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)
<!-- optional: Link to related other tasks on Phabricator. -->
<!-- * https://vyos.dev/Txxxx -->
* https://vyos.dev/T6425
 
## Related PR(s)
<!-- Link here any PRs in other repositories that are required by this PR -->

## Component(s) name
<!-- A rather incomplete list of components: ethernet, wireguard, bgp, mpls, ldp, l2tp, dhcp ... -->
wireless

## Proposed changes
<!--- Describe your changes in detail -->
This bugfix addresses errors in the `data/templates/wifi/hostapd.conf.j2` template file which prevented correct generation of `/run/hostapd/wlanX.conf` when VHT beamforming was configured. There was also no smoketest yet to detect such an error. Two tests have been added to this PR to check for correct generation of `/run/hostapd/wlanX.conf` when VHT beamforming is configured.

## How to test
<!---
Please describe in detail how you tested your changes. Include details of your testing
environment, and the tests you ran. When pasting configs, logs, shell output, backtraces,
and other large chunks of text, surround this text with triple backtics
```
like this
```
-->

1. Configure a WiFi interface `wlanX` using the config below.
2. Commit your changes. 
  - When using `mac80211_hwsim`, prepare to see an error because simulated WiFi NICs do not support VHT beamforming.
4. Check `/run/hostapd/wlanX.conf`
  - Verify this line: `vht_capab=[MAX-MPDU-11454][MAX-A-MPDU-LEN-EXP-3][MU-BEAMFORMER][SU-BEAMFORMEE][SU-BEAMFORMER][BF-ANTENNA-2][SOUNDING-DIMENSION-2]`. 
  - Alternatively use the smoketest `/usr/libexec/vyos/tests/smoke/cli/test_interfaces_wireless.py`.

`show interfaces wireless wlanX`:
```
 address 10.255.5.254/24
 capabilities {
     ht {
         channel-set-width ht20
         channel-set-width ht40+
         channel-set-width ht40-
         dsss-cck-40
         short-gi 20
         short-gi 40
     }
     vht {
         antenna-count 3
         beamform multi-user-beamformer
         beamform single-user-beamformee
         beamform single-user-beamformer
         center-channel-freq {
             freq-1 42
         }
         channel-set-width 1
         max-mpdu 11454
         max-mpdu-exp 3
     }
 }
 channel 36
 country-code de
 description "5GHz 802.11ac"
 mode ac
 security {
     wpa {
         mode wpa2
         passphrase secretpassphrase123
     }
 }
 ssid 80211ac-5GHz
 type access-point
```

## Smoketest result
<!-- Provide the output of the smoketest
```
$ /usr/libexec/vyos/tests/smoke/cli/test_xxx_feature.py
test_01_simple_options (__main__.TestFeature.test_01_simple_options) ... ok
```
-->

```
vyos@vyos:~$ /usr/libexec/vyos/tests/smoke/cli/test_interfaces_wireless.py
test_add_multiple_ip_addresses (__main__.WirelessInterfaceTest.test_add_multiple_ip_addresses) ... ok
test_add_single_ip_address (__main__.WirelessInterfaceTest.test_add_single_ip_address) ... ok
test_dhcp_client_options (__main__.WirelessInterfaceTest.test_dhcp_client_options) ... ok
test_dhcp_disable_interface (__main__.WirelessInterfaceTest.test_dhcp_disable_interface) ... ok
test_dhcp_vrf (__main__.WirelessInterfaceTest.test_dhcp_vrf) ... ok
test_dhcpv6_client_options (__main__.WirelessInterfaceTest.test_dhcpv6_client_options) ... ok
test_dhcpv6_vrf (__main__.WirelessInterfaceTest.test_dhcpv6_vrf) ... ok
test_dhcpv6pd_auto_sla_id (__main__.WirelessInterfaceTest.test_dhcpv6pd_auto_sla_id) ... ok
test_dhcpv6pd_manual_sla_id (__main__.WirelessInterfaceTest.test_dhcpv6pd_manual_sla_id) ... ok
test_interface_description (__main__.WirelessInterfaceTest.test_interface_description) ... ok
test_interface_disable (__main__.WirelessInterfaceTest.test_interface_disable) ... ok
test_interface_ip_options (__main__.WirelessInterfaceTest.test_interface_ip_options) ... ok
test_interface_ipv6_options (__main__.WirelessInterfaceTest.test_interface_ipv6_options) ... skipped 'not supported'
test_interface_mtu (__main__.WirelessInterfaceTest.test_interface_mtu) ... skipped 'not supported'
test_ipv6_link_local_address (__main__.WirelessInterfaceTest.test_ipv6_link_local_address) ... skipped 'not supported'
test_mtu_1200_no_ipv6_interface (__main__.WirelessInterfaceTest.test_mtu_1200_no_ipv6_interface) ... skipped 'not supported'
test_span_mirror (__main__.WirelessInterfaceTest.test_span_mirror) ... skipped 'not supported'
test_vif_8021q_interfaces (__main__.WirelessInterfaceTest.test_vif_8021q_interfaces) ... skipped 'not supported'
test_vif_8021q_lower_up_down (__main__.WirelessInterfaceTest.test_vif_8021q_lower_up_down) ... skipped 'not supported'
test_vif_8021q_mtu_limits (__main__.WirelessInterfaceTest.test_vif_8021q_mtu_limits) ... skipped 'not supported'
test_vif_8021q_qos_change (__main__.WirelessInterfaceTest.test_vif_8021q_qos_change) ... skipped 'not supported'
test_vif_s_8021ad_vlan_interfaces (__main__.WirelessInterfaceTest.test_vif_s_8021ad_vlan_interfaces) ... ok
test_vif_s_protocol_change (__main__.WirelessInterfaceTest.test_vif_s_protocol_change) ... ok
test_wireless_access_point_bridge (__main__.WirelessInterfaceTest.test_wireless_access_point_bridge) ... ok
test_wireless_add_single_ip_address (__main__.WirelessInterfaceTest.test_wireless_add_single_ip_address) ... ok
test_wireless_hostapd_config (__main__.WirelessInterfaceTest.test_wireless_hostapd_config) ... ok
test_wireless_hostapd_vht_mu_beamformer_config (__main__.WirelessInterfaceTest.test_wireless_hostapd_vht_mu_beamformer_config) ... ok
test_wireless_hostapd_vht_su_beamformer_config (__main__.WirelessInterfaceTest.test_wireless_hostapd_vht_su_beamformer_config) ... ok
test_wireless_hostapd_wpa_config (__main__.WirelessInterfaceTest.test_wireless_hostapd_wpa_config) ... ok
test_wireless_security_station_address (__main__.WirelessInterfaceTest.test_wireless_security_station_address) ... FAIL

======================================================================
FAIL: test_wireless_security_station_address (__main__.WirelessInterfaceTest.test_wireless_security_station_address)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/libexec/vyos/tests/smoke/cli/test_interfaces_wireless.py", line 434, in test_wireless_security_station_address
    self.assertTrue(process_named_running('hostapd'))
AssertionError: None is not true

----------------------------------------------------------------------
Ran 30 tests in 184.864s

FAILED (failures=1, skipped=9)
```

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- The entire development process is outlined here: https://docs.vyos.io/en/latest/contributing/development.html -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

alainlamar committed:
wireless: T6425: Fix broken VHT beamforming

alainlamar committed:
wireless: T6425: Add smoketests for VHT beamforming

alainlamar commented:
The smoketest `test_wireless_security_station_address` is still failing randomly. 

I debugged this by making the test write away the generated configuration files just before the failing `self.assertTrue(process_named_running('hostapd'))`. Then, I ran the tests as often until `test_wireless_security_station_address` failed. I reviewed all three files, copied `/tmp/wlan0_station_accept.conf` and `/tmp/wlan0_station_deny.conf` back to `/run/hostapd/` and manually started `hostapd` which worked.

Debugging `/usr/libexec/vyos/tests/smoke/cli/test_interfaces_wireless.py`: 
```
433         import shutil
434         shutil.copyfile('/run/hostapd/wlan0_station_accept.conf', '/tmp/wlan0_station_accept.conf')
435         shutil.copyfile('/run/hostapd/wlan0_station_deny.conf', '/tmp/wlan0_station_deny.conf')
436         shutil.copyfile('/run/hostapd/wlan0.conf', '/tmp/wlan0-smoketest.conf')
437         # Check for running process
438         self.assertTrue(process_named_running('hostapd'))
```

Manual run of `hostapd`:
```
vyos@vyos:~$ sudo /usr/sbin/hostapd -P /run/hostapd/wlan0.pid /tmp/wlan0-smoketest.conf 
wlan0: interface state UNINITIALIZED->COUNTRY_UPDATE
ACS: Automatic channel selection started, this may take a bit
wlan0: interface state COUNTRY_UPDATE->ACS
wlan0: ACS-STARTED 
wlan0: ACS-COMPLETED freq=2412 channel=1
wlan0: interface state ACS->ENABLED
wlan0: AP-ENABLED 
```

@c-po Do you have an idea how to fix this random failing?

c-po commented:
> The smoketest `test_wireless_security_station_address` is still failing randomly.
> ...
> @c-po Do you have an idea how to fix this random failing?

See Slack conversation

c-po committed:
wireless: T6425: adjust to latest country-code changes

Commit 9e22ab6b2a ("wireless: T6318: move country-code to a system wide
configuration") removed the per wifi interface setting for a country-code. This
commit adjust the smoketests to the new design.

c-po commented:
@MergifyIo backport circinus sagitta
